### PR TITLE
Save some matrix-vector products in Sternheimer solver

### PR DESCRIPTION
**Pull request by gkemlin**
Two small things I realized yesterday and that I optimized.
- The Schur complement in the Sternheimer solver was not optimized with two Hamiltonian applications per CG iteration, even when extra arrays were empty. So I changed it to one Hamiltonian application per CG iteration, with additional `ham * ψ_extra` which is required only once. This makes it about twice faster !
- Doing this, I realized that `ham * ψ` with `ψ` an array of size zero was not supported in the optimized version of `DftHamiltonian multiplication`, so I added a test to support that. We did not see this before because we had `ψ_extra' * ham * ψ` with `ψ_extra` empty, which is not optimal at all ;)

### CODE REVIEW

File: src/terms/Hamiltonian.jl [hunk 1]
```diff
@@ -117,6 +117,10 @@ end
                                                                            H::DftHamiltonianBlock,
                                                                            ψ::AbstractArray)
     n_bands = size(ψ, 2)
+    # return empty array if ψ is empty
+    if iszero(n_bands)
+        return zeros(size(ψ,1), 0)
```
[the remark this answers is not in the document]
Comment (gkemlin) [reply src/terms/Hamiltonian.jl:122]: No, this raises an error when trying to create a partition of size 0.

[the remark this answers is not in the document]
Comment (antoine-levitt) [reply src/terms/Hamiltonian.jl]: Ah, annoying. Feels like Iterators.partition(1:0, 0) should work and produce an empty iterator... wanna make a PR to julia? :-p

**Comment by antoine-levitt** [comment]
Or just return Hpsi and don't bother

File: src/terms/Hamiltonian.jl [hunk 2]
```diff
@@ -117,6 +117,10 @@ end
                                                                            H::DftHamiltonianBlock,
                                                                            ψ::AbstractArray)
     n_bands = size(ψ, 2)
+    # return empty array if ψ is empty
+    if iszero(n_bands)
+        return zeros(eltype(ψ), size(ψ,1), 0)
+    end
```
Comment (mfherbst) [inline src/terms/Hamiltonian.jl:123]: I would not bother allocating a new thing:
```suggestion
    iszero(n_bands) && return Hψ  # Nothing to do if ψ empty
```

File: src/response/chi0.jl [hunk 3]
```diff
@@ -295,6 +295,7 @@ end
 
     # compute δψnk band per band
     δψ = zero.(ψ)
+    Hψ_extra = ham * ψ_extra
```
Comment (mfherbst) [inline src/response/chi0.jl:298]: I think I would do this inside the `ik` loop k-point by k-point to save some memory.

File: test/chi0.jl [hunk 4]
```diff
@@ -72,6 +72,17 @@ function test_chi0(testcase; symmetries=false, temperature=0,
         diff_applied_χ0 = apply_χ0(scfres, δV)
         @test norm(diff_findiff - diff_applied_χ0) < testtol
 
+        # Test apply_χ0 without extra bands
+        ψ_occ, occ_occ = DFTK.select_occupied_orbitals(basis,
+                                                       scfres.ψ,
+                                                       scfres.occupation;
+                                                       threshold=scfres.occupation_threshold)
+        ε_occ = [scfres.eigenvalues[ik][1:size(ψk,2)] for (ik, ψk) in enumerate(ψ_occ)]
```
Comment (mfherbst) [inline test/chi0.jl:80]: space after `,`

**Pull request merged**